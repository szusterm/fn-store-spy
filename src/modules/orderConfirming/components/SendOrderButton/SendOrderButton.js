import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import {sendOrder} from '../../actions';

import './styles.scss';

export class SendOrderButton extends Component {
	constructor(props) {
		super(props);

		this.state = {
			sendingOrder: false
		};

		this.sendOrder = this.sendOrder.bind(this);
	}

	async sendOrder() {
		this.setSendingOrderState(true);

		const itemsIds = this.getItemsIds();
		await this.props.sendOrder(itemsIds);

		this.setSendingOrderState(false);
	}

	getItemsIds() {
		const {orderedItems} = this.props;

		const itemsIds = [];
		for (const {_id: orderedItemId} of orderedItems) {
			itemsIds.push(orderedItemId);
		}

		return itemsIds;
	}

	setSendingOrderState(sendingOrder) {
		const newState = Object.assign(
			this.state,
			{
				sendingOrder
			}
		);

		this.setState(newState);
	}

	isOrderEmpty() {
		const {orderedItems} = this.props;

		return (orderedItems.length === 0);
	}

	isOrderTooBig() {
		const {orderedItems, maxItemsInOrder} = this.props;

		return (orderedItems.length > maxItemsInOrder);
	}

	render() {
		const {sendingOrder} = this.state;

		return (
			<button
				className={'send-order-button'}
				disabled={(this.isOrderEmpty() || this.isOrderTooBig() || sendingOrder)}
				onClick={this.sendOrder}
			>
				Send order
			</button>
		);
	}
}

SendOrderButton.propTypes = {
	orderedItems: PropTypes.array, //redux
	maxItemsInOrder: PropTypes.number, //redux
	sendOrder: PropTypes.func //redux
};

const mapStateToProps = (state) => {
	const {items} = state.orderCreator;
	const {maxItemsInOrder} = state.config;

	return {
		orderedItems: items,
		maxItemsInOrder
	};
};

const mapDispatchToProps = {sendOrder};

export default connect(mapStateToProps, mapDispatchToProps)(SendOrderButton);