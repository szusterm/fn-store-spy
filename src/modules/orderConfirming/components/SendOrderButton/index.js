import SendOrderButton from './SendOrderButton';

export default SendOrderButton;