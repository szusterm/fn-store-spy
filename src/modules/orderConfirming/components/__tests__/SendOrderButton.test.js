import React from 'react';
import {shallow} from 'enzyme';

import {SendOrderButton} from '../SendOrderButton/SendOrderButton';

const orderedItemsIds = ['45y5tnr5', '56ug657i'];

const orderedItems = [
	{
		_id: orderedItemsIds[0],
		name: 'Something'
	},
	{
		_id: orderedItemsIds[1],
		name: 'Axe'
	}
];

const setup = (propOverrides) => {
	const props = Object.assign({
		orderedItems, //redux
		maxItemsInOrder: 3, //redux
		sendOrder: () => true, //redux
		closeOrderList: () => true //redux
	}, propOverrides);

	const wrapper = shallow(<SendOrderButton {...props}/>);

	return {props, wrapper};
};

describe('SendOrderButton Component', () => {
	it('renders correctly', () => {
		const {wrapper} = setup();

		expect(wrapper).toMatchSnapshot();
	});

	it('calls redux action to send the order with only items ids array after click', () => {
		const {wrapper, props} = setup({
			orderedItems,
			sendOrder: jest.fn()
		});

		wrapper.find('button').simulate('click');

		expect(props.sendOrder).toHaveBeenCalledWith(orderedItemsIds);
	});

	it('makes disabled, during sending the order', () => {
		const {wrapper} = setup({
			sendOrder: jest.fn(() => Promise.resolve(true))
		});

		expect(wrapper.find('button').props().disabled).toBe(false);

		wrapper.find('button').simulate('click');

		expect(wrapper.find('button').props().disabled).toBe(true);
	});

	it('makes disabled if the order is empty', () => {
		const {wrapper} = setup();

		wrapper.setProps({orderedItems: ['item5']});
		const disabledIfNotEmpty = wrapper.find('button').props().disabled;
		expect(disabledIfNotEmpty).toBe(false);

		wrapper.setProps({orderedItems: []});
		const disabledIfEmpty = wrapper.find('button').props().disabled;
		expect(disabledIfEmpty).toBe(true);
	});

	it('makes disabled, if a number of ordered items is higher than max', () => {
		const {wrapper} = setup({
			maxItemsInOrder: 2
		});

		wrapper.setProps({orderedItems: ['item5']});
		const disabledIfNotHigher = wrapper.find('button').props().disabled;
		expect(disabledIfNotHigher).toBe(false);

		wrapper.setProps({orderedItems: ['item1', 'item2', 'item3']});
		const disabledIfHigher = wrapper.find('button').props().disabled;
		expect(disabledIfHigher).toBe(true);
	});
});