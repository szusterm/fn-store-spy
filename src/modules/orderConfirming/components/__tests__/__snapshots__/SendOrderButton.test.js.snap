// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendOrderButton Component renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SendOrderButton
    closeOrderList={[Function]}
    maxItemsInOrder={3}
    orderedItems={
      Array [
        Object {
          "_id": "45y5tnr5",
          "name": "Something",
        },
        Object {
          "_id": "56ug657i",
          "name": "Axe",
        },
      ]
    }
    sendOrder={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <button
        className="send-order-button"
        disabled={false}
        onClick={[Function]}
      >
        Send order
      </button>,
      "className": "col-9",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Send order",
        "className": "send-order-button",
        "disabled": false,
        "onClick": [Function],
      },
      "ref": null,
      "rendered": "Send order",
      "type": "button",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <button
          className="send-order-button"
          disabled={false}
          onClick={[Function]}
        >
          Send order
        </button>,
        "className": "col-9",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Send order",
          "className": "send-order-button",
          "disabled": false,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Send order",
        "type": "button",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
